src/main/java/core/stack/MateStack.java

```diff
@@ -1,19 +1,50 @@
 package core.stack;
 
+import java.util.EmptyStackException;
+
 public class MateStack<T> {
-    public void push(T value) {
+    private Node<T> tail;
+    private int size;
 
+    public void push(T value) {
+        if (tail == null) {
+            tail = new Node<>(value, null);
+        } else {
+            tail = new Node<>(value, tail);
+        }
+        size++;
     }
 
     public T peek() {
-        return null;
+        checkSize();
+        return tail.value;
     }
 
     public T pop() {
-        return null;
+        checkSize();
+        T element = tail == null ? null : tail.value;
+        tail = tail == null ? null : tail.prev;
```

Comment:
we don't need this check
```suggestion
        tail = tail.prev;
```